lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php

```diff
@@ -548,10 +549,12 @@ private function completeIdGeneratorMapping(ClassMetadataInfo $class): void
     {
         $idGenType = $class->generatorType;
         if ($idGenType === ClassMetadata::GENERATOR_TYPE_AUTO) {
-            if ($this->getTargetPlatform()->prefersSequences()) {
+            if (method_exists(AbstractPlatform::class, 'prefersSequences') && $this->getTargetPlatform()->prefersSequences()) {
```

Comment:
:thinking: I think the logic is wrong… let's not call that method for as long as it exists, but call its replacement (`supportsSequence()` ? It's not clear to me, this method was first mentioned in the codebase in 2008) as soon as possible. If it turns out we can remove the calls entirely, let's do that instead?

Reply:
I was afraid I might change the behavior for DBAL drivers that implement `supportsSequences() === true`, but `prefersSequences() === false`. If you tell me that this is not an issue, I can remove that condition. Please advice.

Reply:
@morozov @beberlei maybe you will know the answer? I don't.

Reply:
There's no DBAL logic that would rely on `prefersSequences()`, it was just a bunch of `true`s and `false`s returned depending on the platform. I have no idea why the ORM would prefer one over the other.

Reply:
As far as I can tell, the only platform that supports sequences but doesn't prefer sequences is SQL Server because it supports identity columns as well. So in theory, a call to `prefersSequences()` could be replaced with `supportsSequences() && !supportsIdentityColumns()`. This may preserve the current behavior but I still don't know what problem this logic solves in the first place.

Reply:
Thank you. I've removed the first `if`. Neither `prefers*()` method is called now and if a driver supports both, identity columns and sequences, identity columns win.

Reply:
> identity columns and sequences, identity columns win.

As a consequence, some PostgreSQL-related tests fail now.

Reply:
Gotta love the tests that make assertions against SQL. Especially the SQL is generated by a different package.

Reply:
> As a consequence, some PostgreSQL-related tests fail now.

This is probably why this "prefers" API existed. PostgreSQL also supports sequences and identity columns but for some reason, the ORM prefers sequences when working with PostgreSQL. Maybe this is because identity columns in PostgreSQL are emulated internally via sequences? But in any event, it doesn't seem the ORM and not even the DBAL concern.

Reply:
So how shall we proceed here? Even if I went back to my previous approach, the Postgres tests would fail again as soon as we upgraded to DBAL 3 where the `prefers*()` methods are gone. My best idea would be to duplicate DBAL's concept of the `*Platform` classes for all features that have been removed from DBAL (see also #8887).

Or… well… to re-add those methods to DBAL, of course. 😓

Reply:
In the first commit (69158cd4f3074fb2a08c53375acd1c29087d0040), it was documented that `IDENTITY` was not supported by PostgreSQL:

> `IDENTITY`: Tells Doctrine to use special identity columns in the database that usually generate a value on insertion of a row (i.e. MySql AUTO_INCREMENT). This strategy does currently not provide full portability. IDENTITY is supported by MySql, Sqlite and MsSql.

The current docs state this is not true anymore:

https://github.com/doctrine/orm/blob/95408cd8e41867c38ac9cc2395404e391b3df389/docs/en/reference/basic-mapping.rst#L377-L381

So IMO it's fine no longer to prefer sequences when working with a recent enough PostgreSQL version, not sure which, but maybe @morozov knows.

Reply:
> So IMO it's fine no longer to prefer sequences when working with a recent enough PostgreSQL version, not sure which, but maybe @morozov knows.

Assuming we simply changed the ID generation strategy that is picked if `@GeneratedValue(strategy="AUTO")` was used on Postgres: Would it be safe for existing applications with a Postgres database to upgrade ORM?

Reply:
:thinking: hmmm… obviously not :disappointed: 

Doing a search leads to https://github.com/doctrine/orm/issues/6429

Reply:
To circle back to what possible solution we would have here, I think the best would be to move the "preference" concept to the ORM (and instantly deprecate it as well as `AUTO`): the ORM should not have a preference because it cannot change, since we cannot know what will be the best choice in the future.

Reply:
> 🤔 hmmm… obviously not, the strategy is probably taken into account when creating a new record in the database right? If the ORM assumes `SERIAL` is used, when a sequence actually is => 💥 😞

Yes, that's my expectation as well.

> Maybe we should deprecate `AUTO`.

That's probably a topic that should be discussed in a dedicated ticket. For this PR, my conclusion is: However we want to proceed with the `AUTO` strategy in the future, the behavior of that strategy for a specific database must not change.

Reply:
I created https://github.com/doctrine/orm/issues/8893 for that separate discussion